From 57b7a52cc6b3675bfbff29ef20a509eadce091ac Mon Sep 17 00:00:00 2001
From: Thomas Freitag <Thomas.Freitag@alfa.de>
Date: Wed, 16 Mar 2016 10:36:22 +0100
Subject: [PATCH] Fix rendering of some broken PDF files

Call constructXRef if necessary xref can't be fetched and the PDF don't use xref streams

Bug #92508
---
 poppler/XRef.cc | 8 +++++++-
 poppler/XRef.h  | 3 ++-
 2 files changed, 9 insertions(+), 2 deletions(-)

--- a/poppler/XRef.cc
+++ b/poppler/XRef.cc
@@ -20,7 +20,7 @@
 // Copyright (C) 2007 Carlos Garcia Campos <carlosgc@gnome.org>
 // Copyright (C) 2009, 2010 Ilya Gorenbein <igorenbein@finjan.com>
 // Copyright (C) 2010 Hib Eris <hib@hiberis.nl>
-// Copyright (C) 2012, 2013 Thomas Freitag <Thomas.Freitag@kabelmail.de>
+// Copyright (C) 2012, 2013, 2016 Thomas Freitag <Thomas.Freitag@kabelmail.de>
 // Copyright (C) 2012, 2013 Fabio D'Urso <fabiodurso@hotmail.it>
 // Copyright (C) 2013 Adrian Johnson <ajohnson@redneon.com>
 // Copyright (C) 2013 Pino Toscano <pino@kde.org>
@@ -297,6 +297,7 @@
   ownerPasswordOk = gFalse;
   rootNum = -1;
   strOwner = gFalse;
+  xrefReconstructed = gFalse;
 }
 
 XRef::XRef() {
@@ -1252,6 +1253,11 @@
   return obj;
 
  err:
+  if (!xRefStream && !xrefReconstructed) {
+    rootNum = -1;
+    constructXRef(&xrefReconstructed);
+    return fetch(num, gen, obj, ++recursion);
+  }
   return obj->initNull();
 }
 
--- a/poppler/XRef.h
+++ b/poppler/XRef.h
@@ -19,7 +19,7 @@
 // Copyright (C) 2007 Carlos Garcia Campos <carlosgc@gnome.org>
 // Copyright (C) 2010 Ilya Gorenbein <igorenbein@finjan.com>
 // Copyright (C) 2010 Hib Eris <hib@hiberis.nl>
-// Copyright (C) 2012, 2013 Thomas Freitag <Thomas.Freitag@kabelmail.de>
+// Copyright (C) 2012, 2013, 2016 Thomas Freitag <Thomas.Freitag@kabelmail.de>
 // Copyright (C) 2012, 2013 Fabio D'Urso <fabiodurso@hotmail.it>
 // Copyright (C) 2013 Adrian Johnson <ajohnson@redneon.com>
 //
@@ -201,6 +201,7 @@
   int rootNum, rootGen;		// catalog dict
   GBool ok;			// true if xref table is valid
   int errCode;			// error code (if <ok> is false)
+  GBool xrefReconstructed;	// marker, true if xref was already reconstructed
   Object trailerDict;		// trailer dictionary
   Goffset *streamEnds;		// 'endstream' positions - only used in
 				//   damaged files
